### DISCUSSION TITLE
Help with Using BEMRosetta for Offshore Wind Platform Hydrodynamics

### DISCUSSION THREAD

**Discussion by pelotonlopezmuro** [BEMRosetta/BEMRosetta#53] [Q&A]
Hi @izabala123 and @KoldoRamirez,

Thank you so much for developing and maintaining BEMRosetta! I'm excited to use it for hydrodynamic modeling of offshore wind platforms. My workflow integrates open-source tools as much as possible:

## Preliminary workflow :
My preliminary workflow looks like this: 
1. **CAD Model Preparation**: Create a geometric model in Solidworks and save as a `.stl` file.  
2. **BEMRosetta**: Import the `.stl` file to generate HAMS input files for wave analysis.  
3. **HAMS**: Analyze wave diffraction and radiation using the files from BEMRosetta.  
4. **OpenFAST**: Use WAMIT-styled output files from HAMS to simulate a floating offshore wind turbine.  

## What I Have Done So Far:
For now, I am working with a simple platform, “HywindSpar,” for which I already have the HAMS input files, so I can check my results. I am attaching `OC3-Hywind-Spar.STL` I recently generated.

1. Loaded the `.STL` file.
![image](https://github.com/user-attachments/assets/9ae0b9e1-ab23-4f02-9074-1048fa3a5730)

2. Updated the mass properties:
   - Center of gravity
   - Mass value
   - Moments of inertia 

![image](https://github.com/user-attachments/assets/a89d5354-d8a8-4777-88fe-046c15b3eba4)

3. Exported the `.pnl` files of "all" the body and the "underwater" body.  

## Where I'm Stuck:
1. **Generating `HullMesh.pnl`, `WaterplaneMesh.pnl` and `Hydrostatic.in`:** 
   - I believe that in the "Waterplane" zone, the "Get Hull" and "Fill Lid" options might help, but I'm unsure how to export the two meshes separately.  
2. **Generating `Hydrostatic.in`:**
3. **Understanding Hydrostatic Stiffness:**
   - I tried the "Hydrostatic Stiffness" tab to generate a `.hst` file, but the output doesn't match WAMIT results. Also, the Inertia tab generates the same file.

## Attachments:
I have attached the `OC3-Hywind-Spar.STL` file and mass properties of the OC3-Hywind-Spar for reference.

[OC3-Hywind-Spar.zip](https://github.com/user-attachments/files/17853883/OC3-Hywind-Spar.zip)

| Property                     | Value                |
|------------------------------|----------------------|
| Mass                         | 7466330.00 kg       |
| Volume                       | 8351.79 m³          |
| Center of Mass (X, Y, Z)     | (0.00, 0.00, -89.92) |
| Moments of Inertia at the center of mass  (Lxx, Lyy, Lzz) | (4229230000.00, 4229230000.00, 164230000.00) |
| Moments of Inertia at the body axis  (Ixx, Iyy, Izz) | (64592993432.00, 64592993432.00, 164230000.00) |

## Questions:
1. What is the correct process for generating `HullMesh.pnl`, `WaterplaneMesh.pnl` and `Hydrostatic.in` files?  
2. How should I handle the mass properties input to ensure accurate results (e.g., moments of inertia at the center of mass or the body axis?)?  
3. Are there additional BEMRosetta features or steps I should explore to improve my workflow?
4. Please make comments or suggestions about my workflow that could make me take advantage of the full BEMRosetta potential. 

Thank you in advance for your guidance and suggestions!

**Comment by izabala123** [comment]
Hi Pelotonlopezmuro

`1. What is the correct process for generating HullMesh.pnl, WaterplaneMesh.pnl and Hydrostatic.in files?`
If you upload the body file and the lid file, and fill the inertia matrix, it is enough
![image](https://github.com/user-attachments/assets/9e763ad6-31f9-4da5-b8e5-80577a4192b1)
You don't need to get the wet part of the mesh, BEMRosetta takes care of that.
To get the lid, if your mesh does have it, y  can extract it with 'Extract lid'. If you don't have it, you can create it with 'Fill lid'.

`2. How should I handle the mass properties input to ensure accurate results (e.g., moments of inertia at the centre of mass or the body axis?)?`
If you don't have it, you can get an approximation using the enclosed menu under any of these hypothesis: mass distributed in volume and mass distributed in surface. You should have to set the mass and the body axis. Remember that to work properly you have to supply the full mesh of the device, not only the wet side.
![image](https://github.com/user-attachments/assets/add50695-57a0-49b6-9f3c-529c7b40d3cb)

`3. Are there additional BEMRosetta features or steps I should explore to improve my workflow?`
The geometry is so simple that you may create that in BEMRosetta, in Vessel Mesh/Edit/Revolution.

`4. Please make comments or suggestions about my workflow that could make me take advantage of the full BEMRosetta potential.`
I don't understand this: `I tried the "Hydrostatic Stiffness" tab to generate a .hst file, but the output doesn't match WAMIT results`
Remember that for OpenFAST the .hst file has to be obtained with mass = 0.

Best regards
Iñaki

**Comment by pelotonlopezmuro** [comment]
Hi @izabala123,

Thank you so much for your extensive answer!

I will follow this discussion :)
As you suggested, In the Vessel Mesh tab, I loaded the body mesh, used the 'Fill lid' tool to create it with the lid mesh, and exported it as `.stl`. I tried the 'Simplify' option to see if that mesh could be improved, but there was no change.

![image](https://github.com/user-attachments/assets/10d3cf28-9918-4527-a61a-cd7c33382135)

Then, I loaded the 'Body' and the 'Lid' mesh in the BEM Solver tab.
![image](https://github.com/user-attachments/assets/2ed2cc35-2aa1-469f-bf5e-ffc97ff8284c)

I am still confused about the **Body Axis**. My interpretation of the coordinate **Body Axis**:  the point about which the rotations roll/pitch/yaw are made? Is this correct?

With that in mind: is the **Body Mass/Inertia Matrix** with respect to the **C. Gravity** or with respect to the **Body Axis**?

After saving and choosing HAMS as the BEM solver, I obtain the the Input files and the `HAMS.bat` among other files.
In the Input folder, I get the `Hydrostatic.in` file, which has the 'Center of Gravity' and 'Body Mass Matrix' I specified, but the 'Hydrostatic Restoring Matrix' is empty (zeros).

However, when I execute `WAMIT_MeshTran.bat` I obtain a new(different) `Hydrostatic.in` with 'center of rotation', 'mass matrix', and 'restoring matrix'. In both files, the 'Center of Gravity' and 'center of rotation' are the **C. Gravity**  specified in BEMRosetta. The 'mass matrix' is not the 'Body Mass Matrix'. It is somehow calculated based on the `mesh.gdf` file, but I cannot guess how. Any hints here?

Also, the 'restoring matrix' is populated and seems to be what should be the 'Hydrostatic Restoring Matrix' in the input file. I suppose I need to copy the 'restoring matrix' into the 'Hydrostatic Restoring Matrix' in the `Hydrostatic.in` file to obtain meaningful results?

I am attaching the 'Body' and the 'Lid' `.stl`, the `mesh.gdf` and the two `Hydrostatic.in` files so you can better understand my questions.
[OC3-Hywind-Questions.zip](https://github.com/user-attachments/files/17912279/OC3-Hywind-Questions.zip)

Finally, once I streamline this process, I'd like to use the cmd line version to automate the procedure, avoiding manual steps and errors. I am exploring the `-paramfile` option. Any advice in this regard?

Thank you again for your guidance!

**Comment by izabala123** [comment]
Hi Pelotonlopezmuro

> My interpretation of the coordinate Body Axis: the point about which the rotations roll/pitch/yaw are made? Is this correct?

Yes

>With that in mind: is the Body Mass/Inertia Matrix with respect to the C. Gravity or with respect to the Body Axis?

It should have to be referred to the Body Axis.

Regarding to Hydrostatic.in I advice you to, after running the HAMS.bat, just load in BEMRosetta the Input/ControlFile.in, and manage it. The BEM solvers have particular behaviours and BEMRosetta tries to handle them all, giving comparable results valid to be used, as it is your case, in OpenFAST.

About -paramfile please tell me exactly what you want to do and I will send you the command line.

Best regards
Iñaki

**Comment by pelotonlopezmuro** [comment]
Hi @izabala123,

> `Regarding to Hydrostatic.in I advice you to, after running the HAMS.bat, just load in BEMRosetta the Input/ControlFile.in, and manage it.

When I load the Input/ControlFile in BEMRosetta, I get the BEM Solver 1. General Tab updated. However, when I load in BEMRosetta, the Input/Hydrostatic.in, I get blanks for the C. Gravity and Body Axis coordinates and just zeros in every matrix in the BEM Solver 1. Bodies Tab. **It looks like the loading of the Hydrostatic.in is not properly working?** _I could report that as an issue._  Then, I would have to manually copy the 'restoring matrix' into the 'Hydrostatic Restoring Matrix' in the Hydrostatic. in.

> About -paramfile please tell me exactly what you want to do and I will send you the command line.

Actually, I wanted to ask **which is the preferred approach, using the command line version or the DLL?**
I tried using the command line version using a paramfile to replicate what I have done so far using the GUI, but I could not do it. I just could use 

```
-mesh
-input ".\OC3-Hywind-Spar.STL"
-setid 0
-gethull
-setid 1
-convert ".\Hull.STL"
-exit
```

to generate the hull `.stl` mesh. By the way, each time the command `-setid <id>` is given, the message `Body active model id is -1` is returned, no matter what `<id>`. It seems that the meshid/bemid has not been updated. _I could report that as an issue._

Conceptually speaking, I would like to:

```
MESH
Load body mesh from "OC3-Hywind-Spar.STL"
Fill lid
Export lid mesh to "Fill_lid.STL"

BEM
Load body BEM from "OC3-Hywind-Spar.STL"
Load lid BEM from "Fill_lid.STL"
Set cg
Set c0
Set body mass/inertia matrix
Set external linear damping matrix
Set additional restoring matrix
Set external quadratic damping matrix

Create HAMS input files in ".\HAMS_INPUTS"

Exit
```

**Do you think it is doable without relying on the GUI? How could I do it?**

Thank you once again!

**Comment by izabala123** [comment]
Hi @pelotonlopezmuro 

Let me a couple of days and I will give you the answer.

Best regards
Iñaki

**Comment by izabala123** [comment]
Hi @pelotonlopezmuro

Please you can try to run 
`bemrosetta_cl -paramfile "file.txt"`

with these text included in 'file,txt'. Replace this sample data with the adequate:
```
-mesh
	-input   ".\OC3-Hywind-Spar.STL"
	-getHull    		# Gets the wet hull from all the mesh
	-convert ".\hull.gdf"
	-fillLid 0.5		# Creates a mesh filling the waterplane lid. Mesh of side approx 0.5 of average
	-convert ".\lid.gdf"
-bem
	-new
	-params
		depth 	 100
		rho	 1025
		g	 9.81
		w 	 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1 1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2 2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3
		headings 0 45 90
	-newBody
	-bodyParams 
		c0		0 0 0
		cg      	0 0 -12
		inertia  		100 0   0   0   0   0
					0   100 0   0   0   0
					0   0   100 0   0   0
					0   0   0   2E5 0   0
					0   0   0   0   2E5 0
					0   0   0   0   0   2E5
		linearDamping		100 0   0   0   0   0
					0   100 0   0   0   0
					0   0   100 0   0   0
					0   0   0   2E5 0   0
					0   0   0   0   2E5 0
					0   0   0   0   0   2E5
		quadraticDamping 	100 0   0   0   0   0
					0   100 0   0   0   0
					0   0   100 0   0   0
					0   0   0   2E5 0   0
					0   0   0   0   2E5 0
					0   0   0   0   0   2E5
		additionalStiffness 	100 0   0   0   0   0
					0   100 0   0   0   0
					0   0   100 0   0   0
					0   0   0   2E5 0   0
					0   0   0   0   2E5 0
					0   0   0   0   0   2E5
		mesh   ".\hull.gdf"
		lid    ".\lid.gdf"
	-saveCase
		folder "%DESKTOP%\My Case"
		solver HAMS
		bin
		symYZ
		symXZ
		numthreads 6
```

Best regards
Iñaki

**Comment by pelotonlopezmuro** [comment]
Hi @izabala123,

Thanks for your reply!

Unfortunately, it is not working on my end. I get the `Error: Unknown argument '-fillLid'`, as you can see below:

```
PS C:\ProgramExecutables\BEMRosetta> .\BEMRosetta_cl.exe -paramfile "file.txt"
BEMRosetta. Copyright (c) 2024. Hydrodynamic coefficients converter for Boundary Element Method solver formats
Version beta 2024112016, release, 64 bits
|**------------------| Loading mesh '.\OC3-Hywind-Spar.STL'
File '.\OC3-Hywind-Spar.STL' loaded
Body id 1 hull is got
Model id 1 saved as '.\hull.gdf'
Error: Unknown argument '-fillLid'

In case of doubt try option -h or -help
```

I cannot find `-fillLid` in the list of mesh options (and similarly, I cannot find the other options you suggested using, e.g. `-saveCase` options ). Could it be a spelling issue? **In which source file should I look to get more insight about options that still need to be documented?**

**Comment by izabala123** [comment]
Hi @pelotonlopezmuro 

Please download the file from [here](https://github.com/BEMRosetta/BEMRosetta/releases/download/2024_December/BEMRosetta_cl.exe).
As your geometry is simple, one additional feature that may be added is to create the meshes automatically in BEMRosetta.

Best regards
Iñaki
